### PR TITLE
KAFKA-12497: Skip periodic offset commits for failed source tasks

### PR DESCRIPTION
**Pull request by C0urante**
[Jira](https://issues.apache.org/jira/browse/KAFKA-12497)

This change serves two purposes:
1. Eliminate unnecessary log messages for offset commit of tasks that don't need to perform offset commits (e.g., a task that has failed and for which all data has been flushed and committed)
2. Stop blocking the offset commit thread unnecessarily for flushes that will never succeed because the task's producer has failed to send a record in the current batch with a non-retriable error

Existing unit tests for the `OffsetStorageWriter` are tweaked to verify the small change made to it. Several new unit tests are added for the `WorkerSourceTask` that cover various cases where offset commits should not be attempted, and some existing tests are modified to cover cases where offset commits either should or should not be attempted.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by C0urante** [comment]
@ncliang @gharris1727 @kpatelatwork @ddasarathan could one or two of you take a look at this when you have time?

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerSourceTask.java
Comment (C0urante) [inline connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerSourceTask.java]: Not required for this PR but important to downgrade this to `TRACE` since it contains record contents and we do not log record contents above that level.

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/SourceTaskOffsetCommitter.java
Comment (kpatelatwork) [inline connect/runtime/src/main/java/org/apache/kafka/connect/runtime/SourceTaskOffsetCommitter.java:109]: @C0urante  overall the PR looks good but somehow I feel this `shouldCommitOffsets` is exposing internal details and may be `commitOffsets` should internally encapsulate this.  

How about we move `shouldCommitOffsets` check as first line in the `commitOffsets` method. and remove the `log.debug("{} Committing offsets", workerTask); `from this method as its redundant and `workerTask .commitOffsets` is already doing it.  Your only challenge would be `return false` is in a lot of places in `workerTask .commitOffsets` but we can may be convert the current method to a private method and add a new commitOffsets wrapper method to make that easy. We will also need to move `log.error("{} Failed to commit offsets", workerTask);` in the wrapper method.  This way all we need to do is call `workerTask.commitOffsets()` in try/catch here.

File: connect/runtime/src/main/java/org/apache/kafka/connect/storage/OffsetStorageWriter.java
Comment (kpatelatwork) [inline connect/runtime/src/main/java/org/apache/kafka/connect/storage/OffsetStorageWriter.java]: should this method be synchronized?  All  other access to `data` seems to be in synchronized methods

File: connect/runtime/src/test/java/org/apache/kafka/connect/runtime/WorkerSourceTaskTest.java
Comment (kpatelatwork) [inline connect/runtime/src/test/java/org/apache/kafka/connect/runtime/WorkerSourceTaskTest.java]: two calls to createWorkerTask , is this a  copy paste or I am missing something :)?

File: connect/runtime/src/test/java/org/apache/kafka/connect/runtime/WorkerSourceTaskTest.java
Comment (kpatelatwork) [inline connect/runtime/src/test/java/org/apache/kafka/connect/runtime/WorkerSourceTaskTest.java]: two calls  to createWorkerTask

**Comment by kpatelatwork** [comment]
@C0urante  overall looks good and very good job on the tests but as I am new to the code, I tried my best to review but I recommend getting a LGTM from one more reviewer also who knows this code better than me to ensure we don't miss something obvious?

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/SourceTaskOffsetCommitter.java
Comment (C0urante) [reply connect/runtime/src/main/java/org/apache/kafka/connect/runtime/SourceTaskOffsetCommitter.java:109]: I initially wanted to do just that, but held off for two reasons:

1. The `SourceTaskOffsetCommitter` log message may be valuable to some users, especially if they've muted the `WorkerSourceTask` namespace or at least raised its level to `INFO`.
2. It's a little easier to test this logic separately by creating a separate method that can be queried on its own. The `boolean` return value of `WorkerSourceTask::commitOffsets` makes it difficult to determine if an offset commit has been attempted and failed, or skipped entirely.

I'm not super attached to either approach; if you have thoughts on why one or the other is better I'm happy to hear them.

File: connect/runtime/src/main/java/org/apache/kafka/connect/storage/OffsetStorageWriter.java
Comment (C0urante) [reply connect/runtime/src/main/java/org/apache/kafka/connect/storage/OffsetStorageWriter.java]: Good catch, thanks!

File: connect/runtime/src/test/java/org/apache/kafka/connect/runtime/WorkerSourceTaskTest.java
Comment (C0urante) [reply connect/runtime/src/test/java/org/apache/kafka/connect/runtime/WorkerSourceTaskTest.java]: Whoops! Thanks for catching this.

File: connect/runtime/src/test/java/org/apache/kafka/connect/runtime/WorkerSourceTaskTest.java
Comment (C0urante) [reply connect/runtime/src/test/java/org/apache/kafka/connect/runtime/WorkerSourceTaskTest.java]: Whoops! Thanks for catching this.

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/SourceTaskOffsetCommitter.java
Comment (kpatelatwork) [reply connect/runtime/src/main/java/org/apache/kafka/connect/runtime/SourceTaskOffsetCommitter.java:109]: I prefer the encapsulation approach but will leave the decision to you as it's a minor nitpick :).

I prefer encapsulation because the caller to `commitOffsets` need not worry about first calling `shouldCommitOffsets` and only if it's true then I should call `commitOffsets`.  Think like we don't need to do  `!ArrayList.isEmpty` check before iterating over it :).

Comment (ncliang) [reply connect/runtime/src/main/java/org/apache/kafka/connect/runtime/SourceTaskOffsetCommitter.java:109]: I tend to agree with @kpatelatwork that the `shouldCommit` method is better encapsulated in `WorkerSourceTask` . There is a call site within `WorkerSourceTask` that could also benefit from the `shouldCommit` check, for instance. https://github.com/apache/kafka/blob/a63e5be4195e97e5b825b5912291144d2d0283a3/connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerSourceTask.java#L266

For the testing aspect, could you make the method visible to tests to do the assertion?

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerSourceTask.java
Comment (ncliang) [inline connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerSourceTask.java]: This implementation differs from the implementation of `recordSend` above when `outstandingMessages` contains records other than the current record. `recordSend` will wait until `outstandingMessages` is empty and `flushing` before `notifyAll` . Is this intentional because we've failed and can't clear the outstanding messages?

**Review by ncliang — COMMENTED** [review]
Thanks @C0urante . I also agree with @kpatelatwork that it is better to encapsulate the should offset commit logic in `WorkerSourceTask` . Otherwise I think generally LGTM.

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/SourceTaskOffsetCommitter.java
Comment (C0urante) [reply connect/runtime/src/main/java/org/apache/kafka/connect/runtime/SourceTaskOffsetCommitter.java:109]: So a few things here:

1. Even if `commitOffsets` is called without the `shouldCommitOffsets` check, if any of the conditions that lead to `shouldCommitOffsets` returning `false` arise in the middle of that call to `commitOffsets`, the offset commit will be prematurely aborted and, at that point, effectively become a no-op. So with the analogy of iterating over a list--this is actually perfectly in line with that example; you don't _have_ to call `List::isEmpty` before iterating since iterating over an empty list is effectively a no-op, but you can choose to if you'd like.
2. We could add a `shouldCommitOffsets` guard against the end-of-life offset commit, but since that only occurs once per task and, as mentioned above, it'll now effectively be a no-op if offset commit isn't possible, I don't think it's really necessary. The points of this PR are to avoid cluttering up the single-threaded source task offset committer and to avoid continually emitting useless and confusing log messages. Gating on a one-time offset commit that only takes place once the task is shut down isn't really necessary for either of these.
3. The `SourceTaskOffsetCommitter::commit` method's "{} Committing offsets" log message is super old and I'm really hesitant to mess with that, even by moving it into a separate class. I know this seems trivial to developers but logs are often the first or second line of defense users have in diagnosing incorrect or unexpected behavior on their workers and I don't want to make that any harder, especially for users with log processing pipelines that parse logging namespaces from messages and allow users to search based on that information.

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerSourceTask.java
Comment (C0urante) [reply connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerSourceTask.java]: > Is this intentional because we've failed and can't clear the outstanding messages?

Yep, exactly. We only need to `notifyAll` here in order to potentially wake up an offset commit thread that's parked [here](https://github.com/C0urante/kafka/blob/01797a5004838d970db21a54bd99a29c7c63f5d4/connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerSourceTask.java#L538) waiting for all outstanding messages to be flushed. If the failed message isn't a part of that batch, or we're not in the middle of flushing at all, no wakeup is necessary.

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/SourceTaskOffsetCommitter.java
Comment (kpatelatwork) [reply connect/runtime/src/main/java/org/apache/kafka/connect/runtime/SourceTaskOffsetCommitter.java:109]: makes sense

**Comment by C0urante** [comment]
@rhauch @tombentley could either of you take a look? It'd be nice to get this merged in time for the upcoming 3.1 release; I know I've seen plenty of people led astray by continued offset commit messages for failed tasks and it'd be great if we could improve their experience.

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/SourceTaskOffsetCommitter.java
Comment (tombentley) [inline connect/runtime/src/main/java/org/apache/kafka/connect/runtime/SourceTaskOffsetCommitter.java]: The method used in the `if` is called `shouldCommitOffsets`, not `areThereOffsetsToBeCommitted`, so is this message accurate, or perhaps the method name is slightly misleading? Maybe amending the message to "...there are no offsets that should be committed"?

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerSourceTask.java
Comment (tombentley) [inline connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerSourceTask.java:201]: Can we add some Javadoc

File: connect/runtime/src/test/java/org/apache/kafka/connect/storage/OffsetStorageWriterTest.java
Comment (tombentley) [inline connect/runtime/src/test/java/org/apache/kafka/connect/storage/OffsetStorageWriterTest.java]: Wouldn't `assertEquals(shouldFlush, writer.willFlush())` and `assertEquals(shouldFlush, writer.beginFlush())` be clearer?

**Review by tombentley — COMMENTED** [review]
@C0urante I've taken a very quick initial look. Given the JIRA is explicitly about logging it would be nice if there were assertions in the tests that directly checked for logging. Could we use a LogCaptureAppender to make such assertions?

File: connect/runtime/src/test/java/org/apache/kafka/connect/storage/OffsetStorageWriterTest.java
Comment (C0urante) [reply connect/runtime/src/test/java/org/apache/kafka/connect/storage/OffsetStorageWriterTest.java]: Yeah, good point. Thought it would make messages clearer for failed assertions but the difference isn't worth the unreadable code.

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/SourceTaskOffsetCommitter.java
Comment (C0urante) [reply connect/runtime/src/main/java/org/apache/kafka/connect/runtime/SourceTaskOffsetCommitter.java]: Sure, the message rewording is fair. Renaming to `areThereOffsetsToBeCommitted` is a little verbose and although it does technically capture the case where the task's producer has failed to send a record with a non-retriable error, I think it may still be a little misleading.

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerSourceTask.java
Comment (C0urante) [reply connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerSourceTask.java:201]: Ack, added.

**Comment by C0urante** [comment]
Thanks @tombentley. Interesting point about the log capture appender but it's a little tricky with the current setup. I've hacked together a prototype that tries to provide realistic coverage by simulating genuine offset commits with calls to `SourceTaskOffsetCommitter::commit` without constructing a full-on integration test; LMK what you think.

If the test cases themselves look acceptable, I suspect we may want to either develop a Connect-specific log capture appender (similar to how Connect has its own embedded testing framework that was largely copied from Streams) or move the Streams-specific `LogCaptureAppender` into a more general-use location, possibly in the clients module. The current prototype's dependency on Streams' testing artifacts isn't desirable as it makes partial builds (which I personally do quite frequently while testing Connect-specific changes) less effective.

**Comment by tombentley** [comment]
> I suspect we may want to either develop a Connect-specific log capture appender (similar to how Connect has its own embedded testing framework that was largely copied from Streams) or move the Streams-specific `LogCaptureAppender` into a more general-use location, possibly in the clients module. The current prototype's dependency on Streams' testing artifacts isn't desirable as it makes partial builds (which I personally do quite frequently while testing Connect-specific changes) less effective.

Agreed about he undesirability of the dependency, and if possible a common LogCaptureAppender would be better than have several copies floating about.

File: connect/runtime/src/test/java/org/apache/kafka/connect/runtime/WorkerSourceTaskTest.java
Comment (tombentley) [inline connect/runtime/src/test/java/org/apache/kafka/connect/runtime/WorkerSourceTaskTest.java]: Any reason not to use multi-resource try-with-resources?

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerSourceTask.java
Comment (tombentley) [inline connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerSourceTask.java]: The "timed out" part of the error is a little misleading now.

**Review by tombentley — COMMENTED** [review]
@C0urante the extra assertions in the test look reasonable to me.

**Comment by C0urante** [comment]
I've rebased onto the latest trunk. I think with the latest changes (especially [these logging improvements](https://github.com/apache/kafka/blob/a3ab7d5b42d02f194a8bdf3047c374f938ca0dad/connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerSourceTask.java#L479-L504)) most of the changes in this PR were made redundant. The only remaining room for improvement IMO is skipping log messages for failed tasks; the other issues (squatting on the source task offset commit thread too long for failed messages to be acknowledged, and misleading users with messages about flushing 0 records) have already been addressed.

I've force-pushed a single commit that brings this PR up to date with the latest trunk; going to push an additional commit later this week that addresses the review comments that have been left on it.

**Comment by C0urante** [comment]
Thanks for the review @tombentley, and apologies for the delay. I've addressed your comments and moved the `LogCaptureAppender` class to the clients module where it can be used by both Connect and Streams.

I hope this looks alright given the changes in https://github.com/apache/kafka/pull/11323 that drastically alter the logic touched on here.

We're probably past the point for 3.1 so this isn't particularly urgent anymore but I believe it'd still be useful to have at some point if you can spare the time. No rush, though.

**Comment by C0urante** [comment]
CC @vvcephei  -- this touches on the `LogCaptureAppender` currently used by Streams; you may want to take a look if you have the time.

File: build.gradle
Comment (vvcephei) [inline build.gradle]: Since you've moved the util, do you still need this dependency?

File: build.gradle
Comment (vvcephei) [inline build.gradle:1277]: I wonder if we can drop this dependency from Streams now. Do you mind checking?

File: checkstyle/import-control.xml
Comment (vvcephei) [inline checkstyle/import-control.xml]: Interesting. We had an import restriction on pulling in a Streams dependency?

Actually, it looks like the util isn't in this package anymore at all, so this is probably not needed.

**Review by vvcephei — APPROVED** [review]
Thanks @C0urante !

I just had some small comments. It would be nice to clean those up before merging, but otherwise it LGTM.

**Review by tombentley — APPROVED** [review]
Thanks @C0urante. I guess we might be able to refactor `kafka.utils.LogCaptureAppender` to use the moved `LogCaptureAppender`, but if so we can do that in another PR.

File: build.gradle
Comment (C0urante) [reply build.gradle]: Good catch, do not need this!

File: checkstyle/import-control.xml
Comment (C0urante) [reply checkstyle/import-control.xml]: Yep, good call, thanks for catching this.

I could be wrong but I think the import control logic requires everything to be explicitly allowed; don't think there's an explicit restriction on the Streams package name.

File: build.gradle
Comment (C0urante) [reply build.gradle:1277]: Ack, will give it a shot. Haven't tested locally, hoping Jenkins can do that for us.

Comment (C0urante) [reply build.gradle:1277]: Jenkins has indeed tested for us and the verdict appears to be that this dependency is still required. I that the way that we're pulling in the clients test library prevents transitive dependencies from being included. We might consider changing `testImplementation project(':clients').sourceSets.test.output` to just `testImplementation project(':clients')` but it looks like all non-core projects use that style; probably best to leave that as-is for now and add the log4j dependency back.

Comment (vvcephei) [reply build.gradle:1277]: Oh, right, I forgot about that. We did that on purpose after something bad got pulled in accidentally. I forget the exact details.

Sorry for the wild goose chase.

**Review by vvcephei — APPROVED** [review]

**Comment by mimaison** [comment]
I see 2 committers approved the changes but this was not merged. 
@C0urante can you rebase to fix the conflicts ? Thanks

**Comment by C0urante** [comment]
This will create other conflicts with https://github.com/apache/kafka/pull/11780. Would it be possible to prioritize that PR and then, once it's merged, fix the conflicts on this one and merge it?

**Comment by mimaison** [comment]
Thanks, that makes sense. I've started reviewing #11780

**Comment by C0urante** [comment]
@mimaison I've finally gotten around to rebasing this one, mind taking another look?

**Comment by mimaison** [comment]
@tombentley @vvcephei Can you take another look?

**Comment by C0urante** [comment]
@mimaison @vvcephei @tombentley I'd like to merge this in order to unblock https://github.com/apache/kafka/pull/12434, which adds a second use case for the `LogCaptureAppender` in the Connect unit tests. Would it be possible to give this another pass sometime next week? Thanks!

**Comment by C0urante** [comment]
It's also been proposed in https://github.com/apache/kafka/pull/12434 that we eliminate the Java `LogCaptureAppender` class entirely and replace it with the [Scala-based version](https://github.com/apache/kafka/blob/1cc1e776f703b180f4bd979e8a551805b3bdc94e/core/src/test/scala/unit/kafka/utils/LogCaptureAppender.scala) that's used in core.

Any thoughts?

**Comment by C0urante** [comment]
Test failures are unrelated; going to merge this as-is. @vvcephei @tombentley Please let me know if you'd like to follow up on this at any point.

File: build.gradle
Comment (ijuma) [reply build.gradle:1277]: Hmm, unless clients has a direct dependency on log4j classes, this shouldn't be here. Note that we have the following:

> testRuntimeOnly libs.slf4jlog4j

Comment (ijuma) [reply build.gradle:1277]: Oh, I see, we moved `LogCaptureAppender` to `clients`. Nevermind.

Comment (C0urante) [reply build.gradle:1277]: Yeah, the dependency is still necessary. While I was in the neighborhood I noticed that this PR accidentally duplicated the Log4j line, though. Filed #12764 to fix.

**Pull request merged**
